### ISSUE TITLE
The code to produce text feature vectors used in DialogueRNN

### ISSUE THREAD

**Issue by zhongpeixiang** [SenticNet/conv-emotion#4]
I compared the text features provided in this repo with my own features extracted by a CNN model described by your DialogueRNN paper. 

Your vectors have larger norm than mine (20 times larger for a vector of size 100). Also, I'm able to effectively learn the model using your vectors. But when I use my vectors, the validation loss keeps increasing and f1 score is fluctuating around 10.

Thanks,
Peixiang

**Comment by soujanyaporia** [comment]
Hi Peixiang,

That is weird! Could you please share your extraction code? 

The feature extraction layer consists of a 1D-CNN. 

Regards,
Soujanya

**Comment by zhongpeixiang** [comment]
> Hi Peixiang,
> 
> That is weird! Could you please share your extraction code?
> 
> The feature extraction layer consists of a 1D-CNN.
> 
> Regards,
> Soujanya

Thank you very much for the quick reply. My CNN model is shown below. I use kernel sizes of 3,4,5, each having 50 filters.

```Python
class CNNExtractor(nn.Module):
    def __init__(self, embed_size, hidden_size, output_size, vocab_size, 
                n_kernels, kernel_sizes, stride=1, padding=0, dropout=0):
        super(CNNExtractor, self).__init__()
        self.embed_size = embed_size
        self.hidden_size = hidden_size
        self.output_size = output_size
        self.vocab_size = vocab_size
        self.n_kernels = n_kernels
        self.kernel_sizes = kernel_sizes
        self.stride = stride
        self.padding = padding
        self.dropout = dropout
        self.hidden = None
        
        # layers
        self.embedding = nn.Embedding(vocab_size, embed_size)
        # torch.backends.cudnn.deterministic=True
        self.convs = nn.ModuleList([nn.Conv2d(1, n_kernels, (kernel, embed_size), stride=stride, padding=padding) for kernel in kernel_sizes])
        conv_size = n_kernels * len(kernel_sizes)
        self.fc1 = nn.Linear(conv_size, hidden_size)
        self.fc2 = nn.Linear(hidden_size, output_size)
        self.dropout_layer = nn.Dropout(dropout)
        
    def conv_and_pool(self, x, conv):
        x = conv(x).squeeze(3)  # (batch_size, n_kernels, max_seq_len-#)
        x = torch.relu(F.max_pool1d(x, x.shape[2]).squeeze(2))
        return x
        
    def forward(self, x, device, extract=False):
        """
        x: tokens: (batch_size, num_turn, max_seq_len), speakers: (batch_size, num_turn)
        """
        # embedding layer
        batch_size, num_turn, max_seq_len = x[0].shape
        x = self.embedding(x[0]) # x: (batch_size, num_turn, max_seq_len, embedding_size)
        x = x.view(batch_size*num_turn, max_seq_len, -1) # x: (batch_size * num_turn, max_seq_len, embedding_size)
        x = x.unsqueeze(dim=1) # x: (batch_size * num_turn, 1, max_seq_len, embedding_size)
        outputs = [self.conv_and_pool(x, conv) for conv in self.convs] # [(batch_size * num_turn, n_kernels),...]
        output = torch.cat(outputs, dim=1) # (batch_size * num_turn, n_kernels * len(kernel_sizes))
        output = self.fc1(output) # (batch_size * num_turn, hidden_size)
        if extract:
            return output
        output = torch.relu(output)
        return self.fc2(self.dropout_layer(output)).view(batch_size, num_turn, -1) # (batch_size, num_turn, output_size)
    
    def init_embedding(self, embedding):
        # fix <pad> and <unk> embedding
        embedding[0].fill_(0)
        embedding[1].fill_(0.5)
        self.embedding.weight.data.copy_(embedding)
```

**Comment by soujanyaporia** [comment]
What embedding do you use?

**Comment by zhongpeixiang** [comment]
> What embedding do you use?

300d Word2Vec embedding trained on 100B Google News

**Comment by soujanyaporia** [comment]
Okay, I will check your code asap. Meanwhile, can you tell me which data are you using? When you trained this CNN, how much f1 are you getting?

**Comment by zhongpeixiang** [comment]
I trained this CNN using the IEMOCAP text data you provided in this repo. I got weighted-F1 around 0.5. 

Is it because the dataset is too small?

**Comment by soujanyaporia** [comment]
Your CNN f-score seems correct as we also reported something similar in our paper. The problem could be in your DialogueRNN clone especially where you use these features in the DialogueRNN.

**Comment by zhongpeixiang** [comment]
> Your CNN f-score seems correct as we also reported something similar in our paper. The problem could be in your DialogueRNN clone especially where you use these features in the DialogueRNN.

You also only use the IEMOCAP dataset for training the CNN feature extractor? By simply replacing my feature vectors with your feature vectors, leaving other things unchanged (labels, speakers, masks), I can boost test F1 from 10 to 62.

**Comment by soujanyaporia** [comment]
I think the labels are not ordered when you saved the CNN-extracted features for train and test data. 

To summarise, you seem to something like this - 

The original data 
``````
Feature_1 Label_1
Feature_2 Label_2
Feature_3 Label_3 

After you extracted features 
``````
Feature_2 Label_1
Feature_3 Label_2
Feature_2 Label_3

**Comment by zhongpeixiang** [comment]
> I think the labels are not ordered when you saved the CNN-extracted features for train and test data.
> 
> To summarise, you seem to something like this -
> 
> The original data
> 
> ```
> Feature_1 Label_1
> Feature_2 Label_2
> Feature_3 Label_3 
> ```
> After you extracted features
> 
> ```
> Feature_1 Label_3
> Feature_2 Label_1
> Feature_3 Label_2
> ```

Thank you very much for the suggestions 💯  

There is a mismatch between extracted features and labels because during training I shuffled the data in-place.

**Comment by soujanyaporia** [comment]
Glad that it has solved your issue 👍
